File: mergesort.kt
/**
 * Dereck Helms
 * CS 4050
 */
import java.io.File
import java.io.IOException

fun mergeSort(alist: List<Int>): Pair<List<Int>, Double> {
    // Perform Merge Sort on a list
    val startTime = System.currentTimeMillis()

    fun merge(left: List<Int>, right: List<Int>): List<Int> {
        val result = mutableListOf<Int>()
        var i = 0
        var j = 0

        while (i < left.size && j < right.size) {
            if (left[i] < right[j]) {
                result.add(left[i])
                i++
            } else {
                result.add(right[j])
                j++
            }
        }

        result.addAll(left.subList(i, left.size))
        result.addAll(right.subList(j, right.size))
        return result
    }

    if (alist.size <= 1) {
        return Pair(alist, (System.currentTimeMillis() - startTime) / 1000.0)
    }

    val mid = alist.size / 2
    val left = mergeSort(alist.subList(0, mid))
    val right = mergeSort(alist.subList(mid, alist.size))

    val result = merge(left.first, right.first)

    val elapsedTime = (System.currentTimeMillis() - startTime) / 1000.0
    return Pair(result, elapsedTime + left.second + right.second)
}

fun testFunction(sortFunction: (List<Int>) -> Pair<List<Int>, Double>, alist: List<Int>) {
    // A test utility function
    val alist2 = alist.toMutableList()
    val res = sortFunction(alist)
    println("Using ${sortFunction::class.simpleName} to sort list: ${alist.take(10)}... w/ ${alist.size} items")
    println("    sort time: ${"%.4f".format(res.second)} seconds")
    alist2.sort()
    println("    sorted correctly?: ${if (res.first == alist2) "y :)" else "n :("}")
}

fun main() {
    try {
        val lines = File("py_vs_X_assign2.txt").readLines()
        val inputList = lines.map { it.toInt() }

        testFunction(::mergeSort, inputList)
    } catch (e: IOException) {
        println("Error: File 'py_vs_X_assign2.txt' not found or cannot be read.")
        System.exit(1)
    }
}

File: insertionsort/mergesort.py
"""
Dereck Helms
CS 4050
"""
import time
import random
import sys

sys.setrecursionlimit(100000)


def mergeSort(alist):
    """Perform Merge Sort on a list."""
    start_time = time.time()

    # Helper function for merging two sorted sublists
    def merge(left, right):
        result = []
        i = j = 0

        while i < len(left) and j < len(right):
            if left[i] < right[j]:
                result.append(left[i])
                i += 1
            else:
                result.append(right[j])
                j += 1

        result.extend(left[i:])
        result.extend(right[j:])
        return result

    if len(alist) <= 1:
        return alist, time.time() - start_time

    mid = len(alist) // 2
    left, left_time = mergeSort(alist[:mid])
    right, right_time = mergeSort(alist[mid:])

    # Merge the sorted sublists
    result = merge(left, right)

    elapsed_time = time.time() - start_time
    return result, elapsed_time + left_time + right_time


if __name__ == '__main__':
    """ Check if the program is being run directly (i.e. not being imported) """

    def testFunction(sort_function, alist):
        """ A test utility function. """
        alist2 = alist.copy()
        res = sort_function(list(alist))
        print(f"Using {sort_function.__name__} to sort list: {alist[:10]}... w/ {len(alist)} items")
        print(f"    sort time: {res[1]:.4f} seconds")
        alist2.sort()
        print(f"    sorted correctly?: {'y :)' if res[0] == alist2 else 'n :('}")

    list1 = [54, 26, 93, 17, 77, 31, 44, 55, 20]  # helpful for early testing
    testFunction(mergeSort, list(list1))

    random.seed(1)
    list2 = list(range(5000))
    random.shuffle(list2)
    testFunction(mergeSort, list(list2))

    list3 = list(range(6000, 1000, -1))
    testFunction(mergeSort, list(list3))
